### PR TITLE
fix: revert docarray explicit transitive dependencies and use extras instead

### PR DESCRIPTION
**Pull request by malandis**
tldr: The docarray [integration
PR](https://github.com/hwchase17/langchain/pull/4483) introduced a pinned dependency to protobuf. This is a docarray dependency, not a langchain dependency. Since this is handled by the docarray dependencies, it is unnecessary here.

Further, as a pinned dependency, this quickly leads to incompatibilities with application code that consumes the library. Much less with a heavily used library like protobuf.

Detail: as we see in the [docarray
integration](https://github.com/hwchase17/langchain/pull/4483/files#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711R81-R83), the transitive dependencies of docarray were also listed as langchain dependencies. This is unnecessary as the docarray project has an appropriate [extras](https://github.com/docarray/docarray/blob/a01a05542d17264b8a164bec783633658deeedb8/pyproject.toml#L70). The docarray project also does not require this _pinned_ version of protobuf, rather [a minimum version](https://github.com/docarray/docarray/blob/a01a05542d17264b8a164bec783633658deeedb8/pyproject.toml#L41). So this pinned version was likely in error.

To fix this, this PR reverts the explicit hnswlib and protobuf dependencies and adds the hnswlib extras install for docarray (which installs hnswlib and protobuf, as originally intended). Because version `0.32.0`
of the docarray hnswlib extras added protobuf, we bump the docarray dependency from `^0.31.0` to `^0.32.0`.

# revert docarray explicit transitive dependencies and use extras instead

## Who can review?

@dev2049 -- reviewed the original PR
@eyurtsev -- bumped the pinned protobuf dependency a few days ago

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -181,10 +179,10 @@ openai = ["openai", "tiktoken"]
 text_helpers = ["chardet"]
 cohere = ["cohere"]
 in_memory_store = ["docarray"]
-hnswlib = ["docarray", "protobuf", "hnswlib"]
+hnswlib = ["docarray"]
```
Comment (hwchase17) [inline pyproject.toml:182]: having `hnswlib` have the docarray dependency only seems weird. should call `docarray`?

Comment (malandis) [reply pyproject.toml:182]: I thought that was strange too but had not opined here. I agree and will change.

Comment (malandis) [reply pyproject.toml]: Added a "docarray" extra in https://github.com/hwchase17/langchain/pull/5015/commits/07a4ded643441e021e0c2634f1d60298f699f2b2

**Review by hwchase17 — APPROVED** [review]
lgtm!

**Comment by eyurtsev** [comment]
This sounds good. I was trying to provide a minimum protobuf constraint before, but poetry locking was failing. If removing it entirely is working then great!
